Signal/src/ViewControllers/ConversationView/ConversationViewController.m

```diff
@@ -2124,6 +2130,206 @@ - (void)didTapFailedOutgoingMessage:(TSOutgoingMessage *)message
+    [self.collectionView scrollToItemAtIndexPath:indexPath
+                                atScrollPosition:UICollectionViewScrollPositionTop
+                                        animated:YES];
+
+    // TODO: Highlight the quoted message?
+}
+
+// This method makes a best effort to determine the best scroll position
+// after loading more interactions when trying to scroll to a quoted
+// message.
+//
+// It prefers to be simple than to correctly handle edge cases like multiple
+// interactions with the same timestamp.
+- (NSUInteger)indexPathRowOfQuotedMessage:(TSQuotedMessage *)quotedMessage
+{
+    OWSAssertIsOnMainThread();
+    OWSAssert(quotedMessage);
+    OWSAssert(quotedMessage.timestamp > 0);
+    OWSAssert(quotedMessage.authorId.length > 0);
+
+    NSUInteger result = 0;
+    for (NSUInteger row = 0; row < self.viewItems.count; row++) {
+        ConversationViewItem *viewItem = self.viewItems[row];
+        if (viewItem.interaction.timestamp > quotedMessage.timestamp) {
+            break;
+        }
+        result = row;
+    }
+    return result;
+}
+
+// If result range is valid, then threadInteractionCount will have a valid value too.
+- (NSRange)findRangeOfQuotedMessage:(TSQuotedMessage *)quotedMessage
+             threadInteractionCount:(NSUInteger *)threadInteractionCount
+{
+    OWSAssertIsOnMainThread();
+    OWSAssert(quotedMessage);
+    OWSAssert(quotedMessage.timestamp > 0);
+    OWSAssert(quotedMessage.authorId.length > 0);
+    OWSAssert(threadInteractionCount);
+
+    *threadInteractionCount = 0;
+
+    // We try to find the "quoted interaction" AND
+    // the range within the mapping that includes it.
+    // NOTE: Since the range _IS NOT_ filtered by author,
+    // and timestamp collisions are possible, it's possible
+    // for:
+    //
+    // * The range to include more than the "quoted interaction".
+    // * The range to be non-empty but NOT include the "quoted interaction",
+    //   although this would be a bug.
+    __block NSRange itemRange;
+    itemRange.location = NSNotFound;
+
+    [self.uiDatabaseConnection readWithBlock:^(YapDatabaseReadTransaction *transaction) {
+        TSInteraction *_Nullable quotedInteraction = [self findInteractionInThreadByTimestamp:quotedMessage.timestamp
+                                                                                     authorId:quotedMessage.authorId
+                                                                                  transaction:transaction];
+        if (!quotedInteraction) {
+            return;
+        }
+
+        YapDatabaseAutoViewTransaction *_Nullable extension =
+            [transaction extension:TSMessageDatabaseViewExtensionName];
+        if (!extension) {
+            OWSFail(@"%@ Couldn't load view.", self.logTag);
+            return;
+        }
+
+        *threadInteractionCount = [extension numberOfItemsInGroup:self.thread.uniqueId];
```

Comment:
Just a nit, but this `threadInteractionCount` seems like an aside to the main focus of this method - finding the range of the interaction. 

If it's just here to avoid opening a second transaction then IMO it would be clearer to have the caller open the transaction, get the `threadInteractionCount` at that stack frame, and then pass the transaction into this method.

No biggie though.

Reply:
No longer needed.